File: CSIT111/Lecture_2/Labs/PenghianAng_Lab2.java
// Name: PengHian Ang
// Tutorial: T02P
// Java version: 17.0.7, 64 bit
//
// Objective: Write a Java program to implement the encryption and the decryption
// Name of the program: Penghian_Ang_Lab2.java
// Declaration: This is my own program

import java.util.Scanner;
import static java.lang.System.out;

public class PenghianAng_Lab2 {
    public static void main(String[] args) {
        // Declare and initialise variables
        Scanner scanner = new Scanner(System.in);
        String CipherService = "";

        // Ask user if they want to encrypt or decrypt
        while (true) {
            out.print("Do you want to encrypt or decrypt?\n1. Encrypt\n2. Decrypt\nEnter (1/2): ");
            String userInput_EncryptOrDecrypt = scanner.nextLine();

            if (userInput_EncryptOrDecrypt.equals("1") || userInput_EncryptOrDecrypt.equals("2")) {
                if (userInput_EncryptOrDecrypt.equals("1")) {
                    out.println("\nYou choosen Encrypt\n");
                    CipherService = userInput_EncryptOrDecrypt;
                } else {
                    out.println("\nYou choosen Decrypt\n");
                    CipherService = userInput_EncryptOrDecrypt;
                }
                break;
            } else {
                out.println("\nInvalid input. Please enter (1/2).");
            }
        }

        // Validate user input if it's an integer or not and also if it's between 1000 and 9999
        int number = 0;
        boolean valid = false;

        // Input validation
        while (!valid) {
            out.print("Enter a 4-digit integer to be encrypted: ");

            String userInput = scanner.nextLine();

            if (userInput.matches("\\d{4}")) {
                number = Integer.parseInt(userInput);
                valid = true;
            } else {
                out.println("Invalid input. Please enter a 4-digit integer.\n");
            }
        }


        // Encryption
        if (CipherService.equals("1")){
            // Step (a): Add 7 to each digit and modulus by 10
            int digit1 = ((number / 1000) + 7) % 10;
            int digit2 = ((number / 100) % 10 + 7) % 10;
            int digit3 = ((number / 10) % 10 + 7) % 10;
            int digit4 = (number % 10 + 7) % 10;

            // Step (b): Swap the digits
            int temp = digit1;
            digit1 = digit3;
            digit3 = temp;

            temp = digit2;
            digit2 = digit4;
            digit4 = temp;

            // Step (c): Form the new integer
            int encryptedNumber = digit1 * 1000 + digit2 * 100 + digit3 * 10 + digit4;

            out.printf("Encrypted interger is %04d%n",encryptedNumber);
        }

        // Decryption
        if (CipherService.equals("2")){
            // Step (a): Add 3 to each digit and modulus by 10
            int digit1 = ((number / 1000) + 3) % 10;
            int digit2 = ((number / 100) % 10 + 3) % 10;
            int digit3 = ((number / 10) % 10 + 3) % 10;
            int digit4 = (number % 10 + 3) % 10;

            // Step (b): Swap the digits
            int temp = digit1;
            digit1 = digit3;
            digit3 = temp;

            temp = digit2;
            digit2 = digit4;
            digit4 = temp;

            // Step (c): Form the new integer
            int encryptedNumber = digit1 * 1000 + digit2 * 100 + digit3 * 10 + digit4;

            out.printf("Decrypted interger is %04d%n",encryptedNumber);
        }
        scanner.close();
    }
}

File: CSIT111/Lecture_2/Labs/PenghianAng_Lab2.py
class PenghianAng_Lab2:
    def main(self):
        # Ask user if they want to encrypt or decrypt
        while True:
            try:
                user_input = int(input("Do you want to encrypt or decrypt?\n1. Encrypt\n2. Decrypt\nEnter (1/2): "))
                if user_input in [1, 2]:
                    if user_input == 1:
                        print("\nYou chose Encrypt\n")
                        break
                    else:
                        print("\nYou chose Decrypt\n")
                        break
                else:
                    print("\nInvalid input. Please enter (1/2).")
            except ValueError:
                print("\nInvalid input. Please enter (1/2).")

        # Validate user input if it's a string with 4 characters
        number = ""
        valid = False

        # Input validation
        while not valid:
            number = input("Enter a 4-digit number: ")
            if len(number) == 4 and number.isdigit():
                valid = True
            else:
                print("\nInvalid input. Please enter a 4-digit number.")
                
        number = int(number)

        # Encryption
        if user_input == 1:
            # Step (a): Add 7 to each digit and modulus by 10
            digit1 = ((number // 1000) + 7) % 10
            digit2 = (((number // 100) % 10) + 7) % 10
            digit3 = (((number // 10) % 10) + 7) % 10
            digit4 = ((number % 10) + 7) % 10

            # Step (b): Swap the first digit
            digit1, digit3 = digit3, digit1
            digit2, digit4 = digit4, digit2

            encrypted_number = digit1 * 1000 + digit2 * 100 + digit3 * 10 + digit4
            print(f"Encrypted integer is {encrypted_number:04d}\n")

        # Decryption
        if user_input == 2:
            # Step (a): Add 3 to each digit and modulus by 10
            digit1 = ((number // 1000) + 3) % 10
            digit2 = (((number // 100) % 10) + 3) % 10
            digit3 = (((number // 10) % 10) + 3) % 10
            digit4 = ((number % 10) + 3) % 10
            
            # Step (b): Swap the first digit
            digit1, digit3 = digit3, digit1
            digit2, digit4 = digit4, digit2

            encrypted_number = digit1 * 1000 + digit2 * 100 + digit3 * 10 + digit4
            print(f"Decrypted integer is {encrypted_number:04d}\n")

lab2 = PenghianAng_Lab2()
lab2.main()